File: src-backend/services/cash/transfer_service.py
import logging

from sqlalchemy.ext.asyncio import AsyncSession

from enums.transaction_type_enum import TransactionType
from core.exceptions import BusinessRuleError, NotFoundError
from models.account import Account
from models.transaction import Transaction
from models.transfer import Transfer
from schemas.cash.transfer_schemas import TransferCreateSchema

logger = logging.getLogger("kasaio")


class TransferService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def create_transfer(self, data: TransferCreateSchema) -> Transfer:
        if data.from_account_id == data.to_account_id:
            raise BusinessRuleError("from_account_id and to_account_id must be different")

        from_account = await self.db.get(Account, data.from_account_id)
        if not from_account:
            raise BusinessRuleError(f"From account not found: id={data.from_account_id}")
        if not from_account.is_active:
            raise BusinessRuleError(f"From account is inactive: id={data.from_account_id}")

        to_account = await self.db.get(Account, data.to_account_id)
        if not to_account:
            raise BusinessRuleError(f"To account not found: id={data.to_account_id}")
        if not to_account.is_active:
            raise BusinessRuleError(f"To account is inactive: id={data.to_account_id}")

        if from_account.currency != to_account.currency and data.exchange_rate is None:
            raise BusinessRuleError(
                "exchange_rate is required when transferring between accounts with different currencies"
            )

        from_tx = Transaction(
            account_id=data.from_account_id,
            type=TransactionType.TRANSFER,
            amount=-abs(data.from_amount),
            currency=from_account.currency,
            date=data.date,
            description=data.description,
            category_id=None,
        )
        to_tx = Transaction(
            account_id=data.to_account_id,
            type=TransactionType.TRANSFER,
            amount=abs(data.to_amount),
            currency=to_account.currency,
            date=data.date,
            description=data.description,
            category_id=None,
        )
        self.db.add(from_tx)
        self.db.add(to_tx)
        await self.db.flush()

        transfer = Transfer(
            from_transaction_id=from_tx.id,
            to_transaction_id=to_tx.id,
            exchange_rate=data.exchange_rate,
        )
        self.db.add(transfer)
        await self.db.commit()
        await self.db.refresh(transfer)
        logger.info(
            f"Transfer created: id={transfer.id} "
            f"from_account={data.from_account_id} to_account={data.to_account_id}"
        )
        return transfer

    async def delete_transfer(self, transfer_id: int) -> None:
        transfer = await self.db.get(Transfer, transfer_id)
        if not transfer:
            raise NotFoundError(f"Transfer not found: id={transfer_id}")

        from_tx = await self.db.get(Transaction, transfer.from_transaction_id)
        to_tx = await self.db.get(Transaction, transfer.to_transaction_id)

        await self.db.delete(transfer)
        if from_tx:
            await self.db.delete(from_tx)
        if to_tx:
            await self.db.delete(to_tx)
        await self.db.commit()
        logger.info(f"Transfer deleted: id={transfer_id}")

File: packages/main/src/services/TransferService.ts
import { AccountRepository } from '../repositories/AccountRepository.js'
import { TransactionRepository } from '../repositories/TransactionRepository.js'
import { TransferRepository } from '../repositories/TransferRepository.js'
import { getDb } from '../db/connection.js'
import {
  NotFoundError,
  BusinessRuleError,
  ErrorCode,
  type Transfer,
  type TransferRequest,
  type Transaction,
  type NewTransaction,
} from '@kasaio/shared'
import { Money } from '../domain/Money.js'
import { ExchangeRate } from '../domain/ExchangeRate.js'

// A transfer is returned with its two transactions so the IPC layer can
// format them without a second lookup.
export interface TransferWithTransactions {
  transfer: Transfer
  fromTransaction: Transaction
  toTransaction: Transaction
}

export class TransferService {
  constructor(
    private repository: TransferRepository,
    private accountRepository: AccountRepository,
    private transactionRepository: TransactionRepository,
  ) {}

  getById(id: number): TransferWithTransactions {
    const transfer = this.repository.findById(id)

    if (!transfer) {
      throw new NotFoundError(ErrorCode.TRANSFER_NOT_FOUND, `Transfer with id ${id} not found.`, { id })
    }

    return {
      transfer,
      fromTransaction: this.transactionRepository.findById(transfer.from_transaction_id)!,
      toTransaction: this.transactionRepository.findById(transfer.to_transaction_id)!,
    }
  }

  create(data: TransferRequest): TransferWithTransactions {
    if (data.from_account_id === data.to_account_id) {
      throw new BusinessRuleError(ErrorCode.SELF_TRANSFER, 'A transfer must be between two different accounts.', { accountId: data.from_account_id })
    }

    const fromAccount = this.accountRepository.findById(data.from_account_id)
    if (!fromAccount) {
      throw new NotFoundError(ErrorCode.ACCOUNT_NOT_FOUND, `Account with id ${data.from_account_id} not found.`, { id: data.from_account_id })
    }
    
    const toAccount = this.accountRepository.findById(data.to_account_id)
    if (!toAccount) {
      throw new NotFoundError(ErrorCode.ACCOUNT_NOT_FOUND, `Account with id ${data.to_account_id} not found.`, { id: data.to_account_id })
    }

    const sameCurrency = fromAccount.currency === toAccount.currency
    if (!sameCurrency && data.exchange_rate == null) {
      throw new BusinessRuleError(ErrorCode.MISSING_EXCHANGE_RATE, `An exchange rate is required to transfer between ${fromAccount.currency} and ${toAccount.currency}.`)
    }

    // Money leaves the from-account and arrives in the to-account's currency
    // (converted when the two currencies differ).
    const rate = sameCurrency
      ? null
      : ExchangeRate.fromDecimal(data.exchange_rate!, fromAccount.currency, toAccount.currency)

    const fromMoney = Money.fromDecimal(data.amount, fromAccount.currency)
    const toMoney = rate ? rate.convert(fromMoney) : Money.fromDecimal(data.amount, toAccount.currency)

    const storedRate = rate ? rate.toStored() : null

    const newFromTransaction: NewTransaction = {
      account_id: fromAccount.id,
      category_id: null,
      transaction_type: 'transfer',
      amount: -Math.abs(fromMoney.toMinorUnits()),
      currency: fromAccount.currency,
      transacted_at: data.transacted_at,
      description: data.description ?? null,
    }
    const newToTransaction: NewTransaction = {
      account_id: toAccount.id,
      category_id: null,
      transaction_type: 'transfer',
      amount: Math.abs(toMoney.toMinorUnits()),
      currency: toAccount.currency,
      transacted_at: data.transacted_at,
      description: data.description ?? null,
    }

    const result = getDb().transaction((): TransferWithTransactions => {
      const fromTransaction = this.transactionRepository.create(newFromTransaction)
      const toTransaction = this.transactionRepository.create(newToTransaction)
      const transfer = this.repository.create({
        from_transaction_id: fromTransaction.id,
        to_transaction_id: toTransaction.id,
        exchange_rate: storedRate,
      })
      return { transfer, fromTransaction, toTransaction }
    })()

    return result
  }

  delete(id: number): void {
    const transfer = this.repository.findById(id)
    if (!transfer) {
      throw new NotFoundError(ErrorCode.TRANSFER_NOT_FOUND, `Transfer with id ${id} not found.`, { id })
    }

    getDb().transaction(() => {
      this.repository.delete(id)
      this.transactionRepository.delete(transfer.from_transaction_id)
      this.transactionRepository.delete(transfer.to_transaction_id)
    })()
  }
}
